### ISSUE TITLE
last_features

### ISSUE THREAD

**Issue by LogSSim** [ultralytics/ultralytics#16872]
### Search before asking

- [X] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

How can I obtain the model last layer features? I tried to get the features, but I just got a Detect type data.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @LogSSim, thank you for reaching out and your interest in Ultralytics 🚀! An Ultralytics engineer will review your query soon.

In the meantime, if this is a feature-related question, we recommend checking the [Docs](https://docs.ultralytics.com) for more information on model layers and usage. If you're dealing with code, providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) could help us understand the issue better.

For additional resources, you can join our community on [Discord](https://ultralytics.com/discord) 🎧 for real-time discussions, or explore [Discourse](https://community.ultralytics.com) for in-depth conversations. Our [Subreddit](https://reddit.com/r/ultralytics) is also a great place for knowledge sharing.

## Upgrade

Ensure you're using the latest `ultralytics` package with all required dependencies in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). You can upgrade using:

```bash
pip install -U ultralytics
```

## Environments

YOLO is supported in several environments with all necessary dependencies preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** or **Amazon** Deep Learning VMs
- **Docker Image**. Check the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

Stay updated with our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule). Green badges indicate all tests are passing across all supported operating systems.

Looking forward to assisting you further! 😊

**Comment by Y-T-G** [comment]
The last layer features are the classification and regression branch outputs. You can get the feature of the second last layer.

```python
model.predict(..., embed=[22])
```

**Comment by LogSSim** [comment]
> 最后一层的特征是分类和回归分支的输出。你可以得到倒数第二层的特征。
> 
> ```python
> model.predict(..., embed=[22])
> ```

Thanks, it works now.

**Comment by glenn-jocher** [comment]
Great to hear it's working for you! If you have any more questions, feel free to ask.

**Comment by LogSSim** [comment]
> 最后一层的特征是分类和回归分支的输出。你可以得到倒数第二层的特征。
> 
> ```python
> model.predict(..., embed=[22])
> ```
![image](https://github.com/user-attachments/assets/e92f7203-26ae-43db-8304-01e421cc66cb)
想请问一下，这个[22]输出后是256，具体代表什么呢？这是倒数第二层，因为我输出print(results[0].masks.shape)为（3,640,640），3应该代表有三个物体，那么这里256具体后面是得到什么呢？想请您解释一下，谢谢

**Comment by LogSSim** [comment]
> Great to hear it's working for you! If you have any more questions, feel free to ask.

May I ask what this [22] output of 256 specifically represents? This is the penultimate layer. Because when I output print(results[0].masks.shape) as (3, 640, 640), 3 should represent three objects. So what exactly does 256 here get later? I would like you to explain it. Thank you.
![image](https://github.com/user-attachments/assets/7d4b260e-1dff-4278-bf84-173ba8db4698)

**Comment by glenn-jocher** [comment]
The `[22]` output with a dimension of 256 represents the feature embeddings from the 22nd (penultimate) layer of the model. These 256 values correspond to the feature vector for each detection, capturing high-level information about the detected objects. This is distinct from the `results[0].masks.shape` output, which represents the segmentation masks for the predicted objects. The `3` in `(3, 640, 640)` corresponds to the number of objects detected.

Let me know if further clarification is needed!

**Comment by LogSSim** [comment]
> The `[22]` output with a dimension of 256 represents the feature embeddings from the 22nd (penultimate) layer of the model. These 256 values correspond to the feature vector for each detection, capturing high-level information about the detected objects. This is distinct from the `results[0].masks.shape` output, which represents the segmentation masks for the predicted objects. The `3` in `(3, 640, 640)` corresponds to the number of objects detected.
> 
> Let me know if further clarification is needed!

OK. If I would like to obtain the segmentation features. What is the number embedded?

**Comment by LogSSim** [comment]
> To obtain the segmentation features, the embedding number corresponds to the specific layer where segmentation features are extracted. Typically, this would be the dedicated segmentation head or a related layer in the model. For precise layers, review the model architecture or inspect layer outputs using tools like PyTorch hooks. Let me know if you'd like further guidance!

![image](https://github.com/user-attachments/assets/f137596c-0c3b-4fa4-ba9e-74623f1799ae)
Alright, I downloaded the pretrained weights for segmentation. How can I access the penultimate layer?

**Comment by glenn-jocher** [comment]
To access the penultimate layer's features in the segmentation model, you can specify the layer index during inference. If you're using `model.predict`, you can pass the `embed` argument with the appropriate layer index, for example:

```python
model.predict(..., embed=[22])  # Replace 22 with the specific index for your model
```

To determine the exact layer index for your case, inspect the model architecture using:

```python
print(model.model)
```

This will help identify the segmentation head or the layer just before the final output. Let me know if you need further clarification!

**Comment by LogSSim** [comment]
> To access the penultimate layer's features in the segmentation model, you can specify the layer index during inference. If you're using `model.predict`, you can pass the `embed` argument with the appropriate layer index, for example:
> 
> ```python
> model.predict(..., embed=[22])  # Replace 22 with the specific index for your model
> ```
> 
> To determine the exact layer index for your case, inspect the model architecture using:
> 
> ```python
> print(model.model)
> ```
> 
> This will help identify the segmentation head or the layer just before the final output. Let me know if you need further clarification!

Okay, if I just want to use the YOLO segmentation model to extract a specific feature, what should I do?

**Comment by glenn-jocher** [comment]
@LogSSim to extract specific features using the YOLO segmentation model, identify the layer of interest by inspecting the model architecture (`print(model.model)`), then specify the desired layer index with the `embed` argument in `model.predict`, like this:

```python
model.predict(..., embed=[layer_index])  # Replace `layer_index` with the target layer
```

Let me know if you need help pinpointing the exact layer!

**Comment by LogSSim** [comment]
> @LogSSim to extract specific features using the YOLO segmentation model, identify the layer of interest by inspecting the model architecture (`print(model.model)`), then specify the desired layer index with the `embed` argument in `model.predict`, like this:
> 
> ```python
> model.predict(..., embed=[layer_index])  # Replace `layer_index` with the target layer
> ```
> 
> Let me know if you need help pinpointing the exact layer!

I've checked it. I want to input a photo into the segmentation model and extract features to concatenate with language features. Which layers should I use for feature extraction? The last one like embed[372]?

**Comment by LogSSim** [comment]
> @LogSSim to extract specific features using the YOLO segmentation model, identify the layer of interest by inspecting the model architecture (`print(model.model)`), then specify the desired layer index with the `embed` argument in `model.predict`, like this:
> 
> ```python
> model.predict(..., embed=[layer_index])  # Replace `layer_index` with the target layer
> ```
> 
> Let me know if you need help pinpointing the exact layer!

When I run model.predict(..., embed=[layer_index]), it returns a tensor with a shape of 768, likely due to pooling. However, I need a tensor in the format (bs, c, h*, w*). What steps should I take? I attempted a solution, but something went wrong:

"""
model = YOLO("pt/yolo11x-seg.pt")  # load an official model
class FeatureExtractor(nn.Module):
    def __init__(self, model, cutoff=22):
        super().__init__()
        self.backbone = nn.Sequential(*list(model.model.children())[:cutoff + 1])

    def forward(self, x):
        return self.backbone(x)

extractor = FeatureExtractor(model, cutoff=22).to("cuda")

x = torch.randn((24, 3, 224, 224)).to("cuda")
with torch.no_grad():
    features = extractor(x)
    print("Feature Shape Before Reduction:", features.shape)

"""

![image](https://github.com/user-attachments/assets/61feaced-4d0a-4cb5-aa8d-21976b633fec)

**Comment by glenn-jocher** [comment]
The issue arises because the `FeatureExtractor` class is incorrectly slicing the `model.model.children()`. YOLO models use composite architectures where layers are not simply sequential. Instead, you can use `forward_hooks` to extract intermediate outputs directly from specific layers. Here's an example approach:

```python
from ultralytics import YOLO
import torch
import torch.nn as nn

# Load YOLO segmentation model
model = YOLO("pt/yolo11x-seg.pt")  

# Hook to capture features from a specific layer
features = []

def hook_fn(module, input, output):
    features.append(output)

# Register hook on the desired layer index
layer_index = 22  # Replace with the desired layer index
target_layer = model.model.model[layer_index]
hook_handle = target_layer.register_forward_hook(hook_fn)

# Input a sample tensor
x = torch.randn((24, 3, 224, 224)).to("cuda")
with torch.no_grad():
    model(x)

# Retrieve and inspect features
print("Feature Shape:", features[0].shape)

# Remove the hook
hook_handle.remove()
```

This method ensures you can capture the desired output tensor in `(bs, c, h, w)` format without modifying the model architecture. Let me know if you encounter further issues!

**Comment by LogSSim** [comment]
> 出现此问题的原因是`FeatureExtractor`类错误地对 进行了切片`model.model.children()`。YOLO 模型使用复合架构，其中的层不是简单的连续层。相反，您可以使用`forward_hooks`直接从特定层提取中间输出。这是一个示例方法：
> 
> ```python
> from ultralytics import YOLO
> import torch
> import torch.nn as nn
> 
> # Load YOLO segmentation model
> model = YOLO("pt/yolo11x-seg.pt")  
> 
> # Hook to capture features from a specific layer
> features = []
> 
> def hook_fn(module, input, output):
>     features.append(output)
> 
> # Register hook on the desired layer index
> layer_index = 22  # Replace with the desired layer index
> target_layer = model.model.model[layer_index]
> hook_handle = target_layer.register_forward_hook(hook_fn)
> 
> # Input a sample tensor
> x = torch.randn((24, 3, 224, 224)).to("cuda")
> with torch.no_grad():
>     model(x)
> 
> # Retrieve and inspect features
> print("Feature Shape:", features[0].shape)
> 
> # Remove the hook
> hook_handle.remove()
> ```
> 
> `(bs, c, h, w)`此方法可确保您无需修改​​模型架构即可以所需格式捕获输出张量。如果您遇到更多问题，请告诉我！
Ok, Thank you very much. I got it.

**Comment by glenn-jocher** [comment]
You're welcome! Glad it helped! Feel free to reach out if you have more questions.
